homeassistant/components/sensor/filter.py

```diff
@@ -0,0 +1,324 @@
+        elif _filter[CONF_FILTER_NAME] == FILTER_NAME_THROTTLE:
+            filters.append(ThrottleFilter(window_size=window_size,
+                                          precision=precision,
+                                          entity=entity_id))
+
+    async_add_devices([SensorFilter(name, entity_id, filters)])
+
+
+class SensorFilter(Entity):
+    """Representation of a Filter Sensor."""
+
+    def __init__(self, name, entity_id, filters):
+        """Initialize the sensor."""
+        self._name = name
+        self._entity = entity_id
+        self._unit_of_measurement = None
+        self._state = None
+        self._filters = filters
+        self._icon = ICON
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Register callbacks."""
+        @callback
+        def filter_sensor_state_listener(entity, old_state, new_state):
+            """Handle device state changes."""
+            self._unit_of_measurement = new_state.attributes.get(
+                ATTR_UNIT_OF_MEASUREMENT)
+            self._icon = new_state.attributes.get(ATTR_ICON, ICON)
+
+            self._state = new_state.state
+
+            if self._state == STATE_UNKNOWN:
+                return
+
+            for filt in self._filters:
+                try:
+                    filtered_state = filt.filter_state(self._state)
+                    _LOGGER.debug("%s(%s, %s) -> %s", filt.name,
+                                  self._entity,
+                                  self._state,
+                                  "skip" if filt.skip else filtered_state)
+                    if filt.skip:
+                        return
+                except ValueError:
+                    _LOGGER.warning("Could not convert state: %s to number",
+                                    self._state)
+                finally:
+                    self._state = filtered_state
+                    filt.states.append(filtered_state)
+
+            self.async_schedule_update_ha_state(True)
+
+        async_track_state_change(
+            self.hass, self._entity, filter_sensor_state_listener)
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit_of_measurement of the device."""
+        return self._unit_of_measurement
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the sensor."""
+        state_attr = {
+            ATTR_ENTITY_ID: self._entity
+        }
+        for filt in self._filters:
```

Comment:
Don't you just want something like this?

```python
attr = {
    "{}_{}".format(filt.name, filt_stat_key): filt_stat_value
    for filt_stat_key, filt_stat_value in filt.stats.items()
    for filt in self._filters
}

attr[ATTR_ENTITY_ID] = self._entity
return attr
```

If we know that `stats()` calls are always used for device_state_attributes, why not prefix it while being generated?

Reply:
Adding the filt.name to all the stats where they are being generated would pollute each filter code.

An option would be in Filter.stats() property, but here it wouldn't be much different then the current case, as the double for would still be present.

Personally I would not expose the stats, but others have pointed out use-case where they want to snag one of the stats attributes. 

Maybe a configuration option? Would that be ok ?

Reply:
MVP has really blown up here he. All of a sudden we have filter chaining, 10s of stats attributes. Duplicate attribute names (if a double filter used) that will override each other